bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,463 @@
+        try {
+            String command = buildShadeCommand("mylink.move.up", targetId);
+            sendCommand(command);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade up: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade up", e);
+        }
+    }
+
+    public void commandShadeDown(String targetId) throws SomfyMyLinkException {
+        try {
+            String command = buildShadeCommand("mylink.move.down", targetId);
+            sendCommand(command);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade down: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade down", e);
+        }
+    }
+
+    public void commandShadeStop(String targetId) throws SomfyMyLinkException {
+        try {
+            String command = buildShadeCommand("mylink.move.stop", targetId);
+            sendCommand(command);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade stop: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade stop", e);
+        }
+    }
+
+    public void commandScene(Integer sceneId) throws SomfyMyLinkException {
+        try {
+            String command = buildSceneCommand("mylink.scene.run", sceneId);
+            sendCommand(command);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding scene run: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding scene run", e);
+        }
+    }
+
+    private void sendCommand(String command) throws SomfyMyLinkException {
+        synchronized (CONNECTION_LOCK) {
+            try {
+                logger.debug("Sending: {}", command);
+                try (Socket socket = getConnection(); OutputStream out = socket.getOutputStream()) {
+                    byte[] sendBuffer = command.getBytes(StandardCharsets.US_ASCII);
+                    // send the command
+                    out.write(sendBuffer, 0, sendBuffer.length);
+                }
+
+                // give time for mylink to process
+                Thread.sleep(CONNECTION_DELAY);
+            } catch (SocketTimeoutException e) {
+                logger.warn("Timeout sending command to mylink: {} Message: {}", command, e.getMessage());
+                throw new SomfyMyLinkException("Timeout sending command to mylink", e);
+            } catch (IOException e) {
+                logger.warn("Problem sending command to mylink: {} Message: {}", command, e.getMessage());
+                throw new SomfyMyLinkException("Problem sending command to mylink", e);
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting after sending command to mylink: {} Message: {}", command,
+                        e.getMessage());
+            }
+        }
+    }
```

Comment:
Because callers of this method might be a UI thread, or a rule thread, you should make sure that you don't block callers with a sleep command, especially when that sleep command is synchronized.

Since it seems that you are trying to throttle the rate at which you send commands through the socket, I suggest that you create a single threaded executor that would handle the command execution for you. Because it is a single thread, you wouldn't need to handle locking and you can perform thread sleeps without worry of blocking other threads. You can look at the bluetooth AM43 binding code as an example of how to create and manage the executor instance.

https://github.com/openhab/openhab-addons/blob/a0f7c4d77df9158bd591d95f8243b66b2ede910d/bundles/org.openhab.binding.bluetooth.am43/src/main/java/org/openhab/binding/bluetooth/am43/internal/AM43Handler.java#L94

In addition, you need to change `sendCommand` to support asynchronous execution. The best way to do this would be to utilize `CompletableFuture`.
```suggestion
    private CompletableFuture<@Nullable Void> sendCommand(String command) {
        CompletableFuture<@Nullable Void> future = new CompletableFuture<>();
        commandExecutor.execute(() -> {
            try {
                logger.debug("Sending: {}", command);
                try (Socket socket = getConnection(); OutputStream out = socket.getOutputStream()) {
                    byte[] sendBuffer = command.getBytes(StandardCharsets.US_ASCII);
                    // send the command
                    out.write(sendBuffer, 0, sendBuffer.length);
                }

                // give time for mylink to process
                Thread.sleep(CONNECTION_DELAY);
            } catch (SocketTimeoutException e) {
                logger.warn("Timeout sending command to mylink: {} Message: {}", command, e.getMessage());
                completableFuture.completeExceptionally(new SomfyMyLinkException("Timeout sending command to mylink", e));
            } catch (IOException e) {
                logger.warn("Problem sending command to mylink: {} Message: {}", command, e.getMessage());
                completableFuture.completeExceptionally(new SomfyMyLinkException("Problem sending command to mylink", e));
            } catch (InterruptedException e) {
                logger.debug("Interrupted while waiting after sending command to mylink: {} Message: {}", command,
                        e.getMessage());
            }
            completableFuture.complete(null);
        });
        return future;
    }
```

Reply:
ok got it. that makes a lot of sense. thank you for showing me around java a bit :)